### PR TITLE
Benchmarking

### PR DESCRIPTION
**Pull request by AnnaWiniwarter**
As previously, please check code and README (in the "Benchmarking" folder) for any bad style violations or something that's inconvenient from a coding perspective and of course anything else that you notice.

### CODE REVIEW

**Comment by AnnaWiniwarter** [comment]
It should run with ixdat v0.2.1 btw. 
I wonder if we should have a separate version for each subproject, since I cannot always guarantee backwards-compatibility...

**Comment by KennethNielsen** [comment]
> It should run with ixdat v0.2.1 btw.
> I wonder if we should have a separate version for each subproject, since I cannot always guarantee backwards-compatibility...

I still think we should keep the version global. I will help with backwards in-compatibility problems.

**Comment by KennethNielsen** [comment]
For GC-stub and plot+fit_gas_exchange this fit looks a little off
Figure_8
.

**Comment by AnnaWiniwarter** [comment]
> > It should run with ixdat v0.2.1 btw.
> > I wonder if we should have a separate version for each subproject, since I cannot always guarantee backwards-compatibility...
> 
> I still think we should keep the version global. I will help with backwards in-compatibility problems.

Ok, if you think that's better. The reason I suggested this is because I will write new contributions always using the newest version of ixdat. How do we then make sure the old contributions also run with the newest version of ixdat?

**Comment by AnnaWiniwarter** [comment]
> For GC-stub and plot+fit_gas_exchange this fit looks a little off Figure_8 .
> 
> EDIT: Ah, I can see that there are more of these. You probably got that covered.

Yes, I am aware. It's because I had some trouble with the MS, so the dataset is not 100% perfect, so the fitting fails in this instance. Might improve this when I record a flawless dataset at some later stage.

File: Benchmarking/Pt_stub/RnD1system_benchmarking_analysis.py [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+# In addition to the above general settings, when using a diffent dataset,
+# the time of when the different measurements occur in the dataset needs to be
+# adjusted in by changing the "tspan" in each section accordingly. To this end
+# it is recommended to use full_data.plot_measurement(tspan=[t1,t2]) after importing
+# and varying t1 and t2 to find the tspans of interest. (see README.md)
+# ----------------------- END OF EDIT SETTINGS ------------------------------
+
+
+# ----------- functions for exponential fitting of decay --------------------
+# for now copy paste from here: https://stackoverflow.com/questions/3938042/fitting-exponential-decay-with-no-initial-guessing
```
Comment (KennethNielsen) [inline Benchmarking/Pt_stub/RnD1system_benchmarking_analysis.py:51]: Maybe update comment

File: Benchmarking/Pt_stub/RnD1system_benchmarking_analysis.py [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+    color : color to use for plotting "signal", for MS signals, standard
+    is selected automatically
+
+    Returns
+    -------
+    t_half, (fig, ax1)
+    """
+    if color is None:
+        if "M" in signal:
+            color = ixdat.plotters.ms_plotter.STANDARD_COLORS[signal]
```
Comment (KennethNielsen) [inline Benchmarking/Pt_stub/RnD1system_benchmarking_analysis.py:85]: also here with `.get`

File: Benchmarking/Pt_stub/RnD1system_benchmarking_analysis.py [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+        else:
+            color = "k"
+    if "M" in signal:
+        ylabel = "MS signal / [A]"
+    else:
+        ylabel = "signal"
+
+    t, sig = data.grab(signal)
```
Comment (KennethNielsen) [inline Benchmarking/Pt_stub/RnD1system_benchmarking_analysis.py:93]: and here about abbreviated names

File: Benchmarking/Pt_stub/RnD1system_benchmarking_analysis.py [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+    ax1.set_xlabel("time / [s]")
+    ax1.set_ylabel(ylabel)
+
+    sig_fit_params = fit_exp_linear(t_to_fit, sig_to_fit, C=0)
+    sig_fit = model_func(t_to_fit, sig_fit_params[0], sig_fit_params[1], C=0)
+
+    ax1.plot(t_to_fit, sig_fit, ":", label="fit")
+    ax1.legend()
+
+    t_half_sig = np.log(2)/-sig_fit_params[1]
```
Comment (KennethNielsen) [inline Benchmarking/Pt_stub/RnD1system_benchmarking_analysis.py:112]: spacing around math operators

File: Benchmarking/Pt_stub/RnD1system_benchmarking_analysis.py [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+
+    gradient_cutoff: defines gradient under which mass signal is "decreasing drastically",
+    depends on the absolute value of the signal -> needs to be chosen individually (a
+    good first guess is -1*((order of magnitude of signal)-1))
+
+    Returns
+    -------
+    t_list: list of times where decay starts
+    """
+    t, m = data.grab(signal)
```
Comment (KennethNielsen) [inline Benchmarking/Pt_stub/RnD1system_benchmarking_analysis.py:139]: abbreviated names

File: Benchmarking/Pt_stub/RnD1system_benchmarking_analysis.py [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+    # now select the timespan where we want to fit the decay curve
+    # the way below is probably not the smartest or most pythonic way, but
+    # it seems to work for this type of data
+    # define a cutoff for select the range where the mass signal is decreasing drastically
+    # choice of this number is a bit arbitrary and doesn't work reliably
+    if gradient_cutoff is None:
+        gradient_cutoff = -np.max(m)/25
```
Comment (KennethNielsen) [inline Benchmarking/Pt_stub/RnD1system_benchmarking_analysis.py:146]: math spacing

File: Benchmarking/Pt_stub/RnD1system_benchmarking_analysis.py [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+        t_list_clean = find_decay_edge(her, signal)  # if automatically generated
+        # list doesn't catch the right times, cutoff value can be selected manually:
+        # gradient_cutoff=-1E-11 (this value works for HER for the tuning used
+        # in this example dataset.)
+        t_half_h2_list = []
+        for time in t_list_clean:
+            t_half_h2, fig = exp_fit_signal(her, signal=signal, tspan=[time, time+5])
+            if SAVE_FIGURES is True:
+                fig[0].savefig("./" + EXP_NAME + "_" + signal +
+                               f"decay_at_{time:.0f}s." + FIGURE_TYPE)
```
Comment (KennethNielsen) [inline Benchmarking/Pt_stub/RnD1system_benchmarking_analysis.py:223]: FIGURE_TYPE double dot

File: Benchmarking/Pt_stub/RnD1system_benchmarking_analysis.py [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+            t_half_he, fig = exp_fit_signal(gas_exchange, signal="M4", tspan=[time, time+5])
+            if SAVE_FIGURES is True:
+                fig[0].savefig("./" + EXP_NAME + f"_M4_decay_at_{time:.0f}s" + FIGURE_TYPE)
+            t_half_list.append(("M4", t_half_he))
+        np.savetxt("./" + EXP_NAME + "_gas_exchange_decay_times.csv", t_half_list,
+                   delimiter=", ", fmt='%s')
+
+    else:
+        raise NameError("WHICH_PART not recognized.")
+
```
Comment (KennethNielsen) [inline Benchmarking/Pt_stub/RnD1system_benchmarking_analysis.py:253]: Here, the same comment as for the other script, about possibly the option to show figures

**Review by KennethNielsen — CHANGES_REQUESTED** [review]
Looks good. There are a few things to consider, main points being:

- Formatting of function argument lists
- Option to show figure
- A double dot in filenames bug. With regards to this on, you can use `git mv` to rename the files also in git, so you wont get a bunch of files deleted and new ones added.

Let me know when you have given it an iteration.

**Comment by KennethNielsen** [comment]
> > > It should run with ixdat v0.2.1 btw.
> > > I wonder if we should have a separate version for each subproject, since I cannot always guarantee backwards-compatibility...
> > 
> > 
> > I still think we should keep the version global. I will help with backwards in-compatibility problems.
> 
> Ok, if you think that's better. The reason I suggested this is because I will write new contributions always using the newest version of ixdat. How do we then make sure the old contributions also run with the newest version of ixdat?

I will setup a system for as soon as a get a little time. It is top of my drawer project list.

File: Benchmarking/GC_stub/RnD1system_benchmarking_analysis_GCstub.py [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+# In addition to the above general settings, when using a diffent dataset,
+# the time of when the different measurements occur in the dataset needs to be
+# adjusted in by changing the "tspan" in each section accordingly. To this end
+# it is recommended to use full_data.plot_measurement(tspan=[t1,t2]) after importing
+# and varying t1 and t2 to find the tspans of interest. (see README.md)
+# ----------------------- END OF EDIT SETTINGS ------------------------------
+
+
+# ----------- functions for exponential fitting of decay --------------------
+# for now copied from here: https://stackoverflow.com/questions/3938042/fitting-exponential-decay-with-no-initial-guessing
```
[the remark this answers is not in the document]
Comment (AnnaWiniwarter) [reply Benchmarking/GC_stub/RnD1system_benchmarking_analysis_GCstub.py:51]: how do I do this properly?

[the remark this answers is not in the document]
Comment (KennethNielsen) [reply Benchmarking/GC_stub/RnD1system_benchmarking_analysis_GCstub.py:51]: Oh, I actually just meant to remove the "for now" as that suggests that it will change :D Fitting an exponential seems like a simple enough task that we should be able to evaluate that the approach you found on SO is valid, so I consider that an ok reference.

File: Benchmarking/GC_stub/RnD1system_benchmarking_analysis_GCstub.py [hunk 10]
```diff
@@ -0,0 +1,259 @@
+# -*- coding: utf-8 -*-
```
[the remark this answers is not in the document]
Comment (AnnaWiniwarter) [reply Benchmarking/GC_stub/RnD1system_benchmarking_analysis_GCstub.py:1]: I learn something every time you go over my code :)

File: Benchmarking/GC_stub/RnD1system_benchmarking_analysis_GCstub.py [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+            t_half_he, fig = exp_fit_signal(gas_exchange, signal="M4", tspan=[time, time+5])
+            if SAVE_FIGURES is True:
+                fig[0].savefig("./" + EXP_NAME + f"_M4_decay_at_{time:.0f}s" + FIGURE_TYPE)
+            t_half_list.append(("M4", t_half_he))
+        np.savetxt("./" + EXP_NAME + "_gas_exchange_decay_times.csv", t_half_list,
+                   delimiter=", ", fmt='%s')
+
+    else:
+        raise NameError("WHICH_PART not recognized.")
+
```
[the remark this answers is not in the document]
Comment (AnnaWiniwarter) [reply Benchmarking/GC_stub/RnD1system_benchmarking_analysis_GCstub.py:254]: When I run with `SAVE_FIGURES = False` I still get the plots, even without `plt.show()` - are you sure it's necessary?

**Comment by AnnaWiniwarter** [comment]
Hi @KennethNielsen, 
Thanks for your review. I implemented most of your suggestions, though I did not touch the docstrings (as I tell myself that no extensive documentation will ever be generated from these functions), and I did not add an option with `plt.show()` as at least for me this is not necessary for the figures to show - does this have something to do with the matplotlib backend that is used? 
If it IS necessary in your case I'm happy to add the option.

**Comment by KennethNielsen** [comment]
> and I did not add an option with plt.show() as at least for me this is not necessary for the figures to show - does this have something to do with the matplotlib backend that is used?
If it IS necessary in your case I'm happy to add the option.

Yes, I think that has something to do with execution environment. I'll leave it up to you, but it **is** required for seeing the figures from another environment.

File: Benchmarking/GC_stub/RnD1system_benchmarking_analysis_GCstub.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (KennethNielsen) [reply Benchmarking/GC_stub/RnD1system_benchmarking_analysis_GCstub.py:273]: I commented on this in the general conversation track

**Review by KennethNielsen — CHANGES_REQUESTED** [review]
So. All the changes look fine. I did however note one new problem while testing whether the figure were shown or not, which is that there is a small issues with the paths still, but easily fixable now that FIGURE_PATH is separately defined. In the beginning of the scripts you have this:
```python
THIS_DIR = Path(__file__).parents[1].resolve()
```

1. which has a number of problems with it (and may have even partially been introduced by me). First one is that it doesn't really do what is promised, because `Path(__file__).parents[1].resolve()` doesn't resolve to the directory of the file, but to the directory one level up, I think because that is what you need the data dir.
2. The other problem is that `parents` is an empty list if the script is run from the command line, because there you don't provide the full path of the file. So in that case really, you should revolve first.

Because I remember that you mentioned that this path stuff is a bit confusing I will demonstrate with a script. Consider this script
```python
from pathlib import Path

print("Path this file", Path(__file__))
print("Parent", Path(__file__).parent)
print("Parents", list(Path(__file__).parents))
```

Run on two different ways, one where the script path is given while being in the directory (usually the case when run from the command line) and one where the script path is given in full:
```bash
❯ python3 pathtest.py
Path this file pathtest.py
Parent .
Parents [PosixPath('.')]

❯ python3 /home/kenneth/tmp/pathtest.py
Path this file /home/kenneth/tmp/pathtest.py
Parent /home/kenneth/tmp
Parents [PosixPath('/home/kenneth/tmp'), PosixPath('/home/kenneth'), PosixPath('/home'), PosixPath('/')]
```

As you can see, unless the full path is given, or the path is resolved, `parents` will only contain one item, which means that `parents[1]` fail, if however you had resolved first, the relative path will become absolute and it will look the same. So, we should do this some other.

Another problem is that while it is great that there is now a configurable `FIGURE_DIR`, it now defaults to `./` which is another way to write the current directory, which means that it has the same problem that the data dir had initially.

To sum up, I suggest to change the code to this (which I believe does what you want, but please check):
 ```python
# ...
THIS_DIR = Path(__file__).parent.resolve()
# ...

# ---------------- EDIT SETTINGS HERE
# ...
DATA_DIR = THIS_DIR.parent / "data"
# ...
FIGURE_DIR = THIS_DIR
# ...
```

where I also bring the definition of THIS_DIR out of the configuration section, since the "directory of the script" isn't really configurable. If you want to have a chat about how it works, let me know.

**Comment by AnnaWiniwarter** [comment]
Thanks for the input. :)
`THIS_DIR = Path(__file__).parents[1].resolve()` 
comes indeed from a suggestion that you made previously, but I wasn't aware of the implications, thanks for the explanation. 

I pushed an iteration now with the changes to the directory definitions, and I also added `plt.show()` at the end - could you please test that this works for you now?

**Comment by KennethNielsen** [comment]
Showing figures with `SAVE_FIGURES=False` does indeed work.

File: Benchmarking/GC_stub/RnD1system_benchmarking_analysis_GCstub.py [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+DATA_SOURCE = "raw"
+# DATA_SOURCE can be "raw" (for importing EC and MS data from Zilien .tsv file),
+# or "ixdat" (for importing ixdat .csv files as generated when using setting "raw")
+# in either case, for plotting the CV only, the biologic CVA file is imported
+# automatically
+WHICH_PART = "plot_CVs"
+# WHICH_PART can be "plot_CVs", "plot+fit_HER", "plot+fit_gas_exchange"
+
+SAVE_FIGURES = True
+# SAVE_FIGURES set to True: figures will be saved automatically.
```
Comment (KennethNielsen) [inline Benchmarking/GC_stub/RnD1system_benchmarking_analysis_GCstub.py:33]: Please mention here, that setting it to False will show the images instead.

File: Benchmarking/GC_stub/RnD1system_benchmarking_analysis_GCstub.py [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+FIGURE_TYPE = ".png"
+# FIGURE_TYPE can be any format matplotlib accepts for saving figures, eg.
+# ".png", ".svg", ".pdf" etc.
+FIGURE_DIR = "./"
```
Comment (KennethNielsen) [inline Benchmarking/GC_stub/RnD1system_benchmarking_analysis_GCstub.py:37]: FIGURE_DIR is now double defined. Please delete this one.

File: Benchmarking/Pt_stub/RnD1system_benchmarking_analysis.py [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+DATA_SOURCE = "ixdat"
+# DATA_SOURCE can be "raw" (for importing EC and MS data from Zilien .tsv file),
+# or "ixdat" (for importing ixdat .csv files)
+# in either case, for plotting the CV only, the biologic CVA file is imported
+# automatically
+WHICH_PART = "plot_CVs"
+# WHICH_PART can be "plot_CVs", "plot+fit_HER", "plot+fit_gas_exchange"
+
+SAVE_FIGURES = True
+# SAVE_FIGURES set to True: figures will be saved automatically.
```
Comment (KennethNielsen) [inline Benchmarking/Pt_stub/RnD1system_benchmarking_analysis.py:32]: Please mention here that setting it to False will show the figures instead

**Review by KennethNielsen — CHANGES_REQUESTED** [review]
2 very small details left. Just let me know when fixed and I'll merge.

File: Benchmarking/GC_stub/RnD1system_benchmarking_analysis_GCstub.py (hunk 13 above)
Comment (AnnaWiniwarter) [reply Benchmarking/GC_stub/RnD1system_benchmarking_analysis_GCstub.py:37]: oops...

**Comment by AnnaWiniwarter** [comment]
> 2 very small details left. Just let me know when fixed and I'll merge.

done :)

**Review by KennethNielsen — APPROVED** [review]
